src/Features/Core/Portable/RemoveUnusedMembers/AvoidUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,199 @@
+
+                compilationStartContext.RegisterSymbolAction(AnalyzeSymbolDeclaration, SymbolKind.Method, SymbolKind.Field, SymbolKind.Property, SymbolKind.Event);
+
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    symbolStartContext.RegisterOperationAction(AnalyzeMemberReferenceOperation, OperationKind.FieldReference, OperationKind.MethodReference, OperationKind.PropertyReference, OperationKind.EventReference);
+                    symbolStartContext.RegisterOperationAction(AnalyzeInvocationOperation, OperationKind.Invocation);
+                    symbolStartContext.RegisterSymbolEndAction(OnSymbolEnd);
+                }, SymbolKind.NamedType);
+
+                return;
+
+                // Local functions.
+                void AnalyzeSymbolDeclaration(SymbolAnalysisContext symbolContext)
+                {
+                    if (IsCandidateSymbol(symbolContext.Symbol))
+                    {
+                        // Initialize unused state to 'true'.
+                        symbolValueUsageStateMap.GetOrAdd(symbolContext.Symbol, valueFactory: s => ValueUsageInfo.None);
+                    }
+                }
+
+                void OnSymbolUsage(ISymbol memberSymbol, ValueUsageInfo usageInfo)
+                {
+                    Debug.Assert(IsCandidateSymbol(memberSymbol));
+
+                    // Update the usage info for the memberSymbol
+                    symbolValueUsageStateMap.AddOrUpdate(
+                        memberSymbol,
+                        addValue: usageInfo,
+                        updateValueFactory: (sym, currentUsageInfo) => currentUsageInfo | usageInfo);
+                }
+
+                void AnalyzeMemberReferenceOperation(OperationAnalysisContext operationContext)
+                {
+                    var memberReference = (IMemberReferenceOperation)operationContext.Operation;
+                    if (IsCandidateSymbol(memberReference.Member))
+                    {
+                        // Get the value usage info.
+                        var valueUsageInfo = memberReference.GetValueUsageInfo();
+
+                        // Special case for VB
+                        // Compound assigment is a statement in VB that does not return a value.
+                        // So, we treat this usage as a write-only usage.
+                        if (memberReference.Language == LanguageNames.VisualBasic &&
+                            memberReference.Parent is ICompoundAssignmentOperation compoundAssignment &&
+                            compoundAssignment.Target == memberReference)
+                        {
+                            valueUsageInfo = ValueUsageInfo.Write;
+                        }
+
+                        OnSymbolUsage(memberReference.Member, valueUsageInfo);
+                    }
+                }
```

Comment:
I'm concerned here because it's not clear to me how IOp encodes error scenarios.  With SemanticModel we have things like "Symbol" and "CandidateSymbols".  With a feature like this, we'd want to be conservative and assume that if we see a Symbol as a CnadidateSymbol that it is used and should not be eligible for removal.  it's unclear to me if this approach ensures that.

Reply:
I don't expect users to use a code fix that removes unused code in presence of errors, and even if they do it is extremely unlikely to be a FixAll. So, I presume it is highly unlikely for the fix to unintentionally break them. Regardless, your concern is not specific to this feature, but IOperation analyzers in general, and I feel we should wait for user feedback before attempting to detect presence of errors and start bailing out.

---
In reply to: [212731651](https://github.com/dotnet/roslyn/pull/29511#discussion_r212731651) [](ancestors = 212731651)

Reply:
> I don't expect users to use a code fix that removes unused code in presence of errors

I highly disagree with that.  :)   

For one thing, it presumes the user is even aware that there *are* errors.  As a user, I do not have my error list even open most of the time.  So i would trust that faded/suggested code was something i could trust.  :)  

>  and I feel we should wait for user feedback before attempting to detect presence of errors and start bailing out.

Note that we have had feedback here in the past for other features.  I'm fairly certain we have a bug *right now* (i'm just too lazy to look) that this happens with 'usings'.  :)

> Regardless, your concern is not specific to this feature, but IOperation analyzers in general,

No... :)   To be clear, it's specific to the user experience.  This is a feature (and i don't care how it's implemented), and it's one taht could have a poor experience for the user *regardless* of fix vs fix-all.  That's because instead of leniently assuming the worst when we have errors, we conservatively assume the best.  i.e. we assume we can just ignore the error information if the compiler doesn't exactly know what's going on.  Now, that's happens to be because this impl decided to sit on top of IOp.  But tha'ts an implementation detail.  The critique would hold if you'd written this the same way on SemanticModel :)

Reply:
I don't feel any feature that relies on presence or lack of symbol references can work perfectly in presence of errors - pretty sure there are cases where SemanticModel/CandidateSymbols returns empty for identifiers whose text matches some symbol due to parser state where the text might just be parsed as trivia, when the user really intended that to be reference to the symbol.

By the way, I think there is a simple way to handle your concern - just register for invalid operations, and let the symbol end bail out completely in presence of even a single invalid operation. I would like to hear feedback from other reviewers before making such a change, as my preference is wait for user feedback before adding such a check.

---
In reply to: [212780470](https://github.com/dotnet/roslyn/pull/29511#discussion_r212780470) [](ancestors = 212780470)

Reply:
> By the way, I think there is a simple way to handle your concern - just register for invalid operations, and let the symbol end bail out completely in presence of even a single invalid operation.

I would prefer that.  It would also match our thoughts on unused-usings.  That they should not even be reported as such if there is a single error in the file.

> I would like to hear feedback from other reviewers before making such a change, as my preference is wait for user feedback before adding such a check.

To be clear: we have already gotten this feedback from users in the past.  If i'm not mistaken we have an outstanding *existing* piece of feedback about a regression we have here already with remove-unused-usings.  :)

The feedback has been: don't break my code by removing things i'm using.  Introducing new removal feature and then just waiting for the same feedback to appear is not advisable IMO given this precedent we already have.

--

Also, i assume at some point that this will become part of CodeCleanup.  This makes it even more critical that we not mess things up.  If people turn this on, and then format, and we start removing things just because they formatted right in the middle of an error, that's just bad news :)

Reply:
Can we take this to be discussed at a design meeting? :)

Reply:
Now tracked with [https://github.com/dotnet/roslyn/issues/29519](https://github.com/dotnet/roslyn/issues/29519), which I hope to bring up at Monday's design meeting.

---
In reply to: [212788605](https://github.com/dotnet/roslyn/pull/29511#discussion_r212788605) [](ancestors = 212788605)